### ISSUE TITLE
3 PHP testcases segfault with PHP 8.0.x

### ISSUE THREAD

**Issue by jplesnik** [swig/swig#2383]
I tried to rebuild SWIG 4.1.0 on CentOS Stream 9 machine and the PHP tests segfault.

```
checking php testcase virtual_poly
checking php testcase virtual_vs_nonvirtual_base (with run test)
make[2]: *** [../../../Examples/Makefile:1074: php_run] Segmentation fault (core dumped)
make[1]: *** [Makefile:53: virtual_poly.cpptest] Error 2
```

Complete build log is [here](https://download.copr.fedorainfracloud.org/results/jplesnik/swig-c9s/centos-stream-9-x86_64/04873254-swig/builder-live.log.gz)

**Comment by ojwb** [comment]
I can't reproduce this on x86-64 debian with PHP 8.0.8, which is the closest to your configuration I have to hand (seems Debian switch to PHP 8.1 after 8.0.8 so nothing newer was packaged).

Can you get a backtrace for where it segfaults?  You should be able to by running just this testcase like so:

```
make -C Examples/test-suite/php virtual_vs_nonvirtual_base.cpptest RUNTOOL='gdb --args'
```

Then `run` at the `(gdb) ` prompt, and `bt` once it segfaults to get a backtrace.  You'll want to have debugging data for PHP installed to get a more useful backtrace.

**Comment by ojwb** [comment]
BTW, it would be helpful to say what commit each of your "4.1.0" bug reports correspond to as 4.1.0 hasn't been released yet.

I tested with eaf468e9e8d3ef17f397ef6bc5e73807a2bf3594.

**Comment by jplesnik** [comment]
@ojwb I tested with the same commit as you. 

These tests segfault with PHP 8.0.20:
[prefix.cpptest](https://jplesnik.fedorapeople.org/swig-4.1.0/bt-prefix)
[director_using_member_scopes.cpptest](https://jplesnik.fedorapeople.org/swig-4.1.0/bt-directory)
[virtual_poly.cpptest](https://jplesnik.fedorapeople.org/swig-4.1.0/bt-virtual)

**Comment by ojwb** [comment]
Thanks, I see segfaults from those 3 testcases too.  Will investigate.

Does `virtual_vs_nonvirtual_base` fail for you individually?  Maybe stdout and stderr lines are not interleaved in the order they were actually written and the segfault wasn't from the testcase it appeared to be from...

**Comment by ojwb** [comment]
We are meant to have CI for PHP 8.0, so I looked at why that hadn't detected this problem.

Turns out we were relying on the default PHP version in the CI image being 8.0, whereas it's actually 8.1 currently so we were testing that twice and not testing 8.0 at all.

I've pushed a fix for that and also added testing of PHP 8.2.

**Comment by ojwb** [comment]
PHP 8.0 now fails in CI (8.2 also fails, but for entirely different reasons from a quick look).

For the `prefix` testcase, the backtrace shows this is happening inside PHP when we register the `ProjectBar` subclass.  It seems to be looking up the `ProjectFoo` base class which we already registered.

I tried `git bisect` on the `prefix` testcase which points to d43f28a6664561b0952722d51702d5e99a806918 but that's just where `prefix_runme.php` was added so a red herring.

`director_using_member_scopes.i` was actually added *after* that commit, but initially worked, so that seems to suggest there isn't a simple common cause here.
